sys/kern/vfs_vnops.c

```diff
@@ -549,12 +549,18 @@ vn_rdwr(enum uio_rw rw, struct vnode *vp, void *base, int len, off_t offset,
 		return (EINVAL);
 	auio.uio_iov = &aiov;
 	auio.uio_iovcnt = 1;
-#if __has_feature(capabilities) && !defined(__CHERI_PURE_CAPABILITY__)
+#ifdef __CHERI_PURE_CAPABILITY__
+	IOVEC_INIT_C(&aiov, base, len);
+#elif __has_feature(capabilities)
 	if (segflg == UIO_USERSPACE)
-		IOVEC_INIT_C(&aiov, __USER_CAP(base, len), len);
+		IOVEC_INIT_C(&aiov,
+		    cheri_capability_build_user_data(CHERI_CAP_USER_DATA_PERMS,
+			(vaddr_t)base, len, 0), len);
 	else
-#endif
 		IOVEC_INIT(&aiov, base, len);
+#else
+	IOVEC_INIT(&aiov, base, len);
+#endif
```

Comment:
```C
#ifdef __CHERI_PURE_CAPABILITY__
	IOVEC_INIT_C(&aiov, base, len);
#else
#if __has_feature(capabilities)
	if (segflg == UIO_USERSPACE)
		IOVEC_INIT_C(&aiov,
		    cheri_capability_build_user_data(CHERI_CAP_USER_DATA_PERMS,
			(vaddr_t)base, len, 0), len);
	else
#endif
		IOVEC_INIT(&aiov, base, len);
#endif
```
is a bit nicer, though why is IOVEC_INIT not the same as IOVEC_INIT_C for a purecap kernel? If it were you could instead have the even nicer:
```C
#if __has_feature(capabilities) && !defined(__CHERI_PURE_CAPABILITY__)
	if (segflg == UIO_USERSPACE)
		IOVEC_INIT_C(&aiov,
		    cheri_capability_build_user_data(CHERI_CAP_USER_DATA_PERMS,
			(vaddr_t)base, len, 0), len);
	else
#endif
		IOVEC_INIT(&aiov, base, len);
```

Reply:
Yeah, I agree that it is ugly, in general the IOVEC initializers are almost equivalent for a purecap kernel. The issue in this case is that `IOVEC_INIT` has an assertion on the address that checks that it is not a user address as in general we always want user addresses to be handled as capabilities at this stage. The assertion remains for the purecap kernel `IOVEC_INIT` but in this specific instance we are providing a capability to userland there, so it would fail the assertion.
This is the only instance where we get this issue with iovec.

Reply:
That assertion only makes sense for a hybrid kernel.

Reply:
`PTR2CAP` is used in many places where we are not supposed to get user capabilities in the purecap kernel as well though, that is why I did not ifdef it to a no-op. However I can probably completely alias `IOVEC_INIT` and `IOVEC_INIT_C` for a purecap kernel though, as the distinction is relevant only in hybrid.